### ISSUE TITLE
Problem with badges in Ventura 13.6.7

### ISSUE THREAD

**Issue by Cocoader** [milke/GitFinder-Issues#126]
There seems to be a problem with activating the finder extension (no badges visible) in 13.6.7 (22G720).
It is similar to the open topic #124.
I tried everything in your "GitFinder is running, but I don't see icon badges nor git menu commands?" but it doesn't solve it.

![Screenshot 2024-06-07 at 00 59 04](https://github.com/milke/GitFinder-Issues/assets/13892544/1a241c32-e3d2-4b65-88ca-2cecb810f6a5)
Extensions / Added extensions:
![Screenshot 2024-06-07 at 01 03 01](https://github.com/milke/GitFinder-Issues/assets/13892544/a57f89b4-b9c4-4374-9e6f-1e706fad51ca)
Extensions / Finder extensions:
![Screenshot 2024-06-07 at 01 24 39](https://github.com/milke/GitFinder-Issues/assets/13892544/95e39d02-a283-4193-8f2b-2444c3139238)
Context menu:
![Screenshot 2024-06-07 at 01 25 42](https://github.com/milke/GitFinder-Issues/assets/13892544/4b183791-88fe-4886-b98d-5623be54bdd5)

**Comment by milke** [comment]
So, you see the GitFinder related contextual menu items, but you don't see GitFinder related icon badges?

**Comment by Cocoader** [comment]
Correct

**Comment by milke** [comment]
I tried to reproduce your problem with exactly the same setup (Ventura 13.6.7, Dropbox, OneDrive), but I'm not able to.  On some relaunches GitFinder takes precedence and shows its icon badges immediately, but on some relaunches, when either Dropbox or OneDrive take precedence, unchecking their Finder extensions in System Preferences enables GitFinder to show its icons. Once that happens, I can "safely" check back on both Dropbox and OneDrive without affecting GitFinder.

If you disable other Finder extensions, what happens if you select in Finder some git repository file (either the root of the repository folder of any file inside it's hierarchy)? Also,, what happens if you disable (uncheck) both Dropbox and OneDrive and then logout and log back in (without enabling the two)? Does GitFinder shows its icon badges?

**Comment by Cocoader** [comment]
Thank you for trying to reproduce this issue.

I did disable Dropbox and OneDrive(without a restart) and browsed a git repository but didn’t see badges on any level.
I also tried it with a logout / login and also a restart. I haven't seen the badges once - not even intermittently.

Is it correct that GitFinder shows only in “Extensions / Added extensions” and not in “Extensions / Finder extensions” like Dropbox or OneDrive?

**Comment by milke** [comment]
> Is it correct that GitFinder shows only in “Extensions / Added extensions” and not in “Extensions / Finder extensions” like Dropbox or OneDrive?

I think we should't draw ay conclusions from what System Settings application displays in _Extensions / Finder extensions_. Unfortunately, that applications i still quite buggy and it's very inconsistent when displaying _Extensions / Finder extensions_. For example, on Ventura (13.6.7, the one I used to try to reproduce your issue), I have four FinderSync extensions. All four are displayed in _Extensions / Added extensions_ with _Finder extensions_ checkmark, like here:
<img width="827" alt="Screenshot 2024-06-13 at 13 11 43" src="https://github.com/milke/GitFinder-Issues/assets/291305/abc72a8a-da1a-4e99-ad20-c7ea32e6fa38">
However. only _One Drive_ is listed in _Extensions / Finder extensions_:
<img width="827" alt="Screenshot 2024-06-13 at 13 11 53" src="https://github.com/milke/GitFinder-Issues/assets/291305/cb3f9194-4ebe-486d-825b-690ed531dc47">
On my main system Sonoma (14.5), I see all four in _Extensions / Added extensions_, but none is listed in _Extensions / Finder extensions_ (not even _One Drive_), that category doesn't exist at all, like you can see here:
<img width="827" alt="Screenshot 2024-06-13 at 14 53 38" src="https://github.com/milke/GitFinder-Issues/assets/291305/ce5b31ab-ade6-4b29-a7be-f96892ca8bfd">
… contrary to this on Ventura (13.6.7):
<img width="827" alt="Screenshot 2024-06-13 at 13 14 12" src="https://github.com/milke/GitFinder-Issues/assets/291305/c26a9e02-9c1f-4b14-91fd-5ef64e2a243f">
So in general, you shouldn't really pay attention to it. The fact you can see GitFinder's menu items is enough to know it functions properly. But at the moment, I don't really have any idea why you don't see icon badges.

Can you please create a temporary test user and try under that fresh account? If that doesn't work as well, then I assume I'll have to create some special test version for you, with some logging, to try to figure out what's going on.

In my defense, I'd just like to add I've filed a bug report at Apple about this badges collision, with a request to solve it somehow, maybe introducing some king of setting for the user to decide which of installed extensions will take precedence if more of them monitor the same folder. I did it a long time, when I noticed it for the first time. Unfortunately, nothing happened regarding that. If you're interested in more technical explanation what actually happens during this collision, you can find it [here](https://github.com/milke/GitFinder-Issues/issues/123#issuecomment-1763062003).

**Comment by milke** [comment]
> I was too focused on the applications in **"Extensions / Finder extensions"** _Dropbox_ and _OneDrive_ that actually create badges on their own.
> It actually was "Hookmark" in **"Extensions / Added extensions"** that has Finder extensions and doesn't show up in **"Extensions / Finder extensions"**.

Yeah, Apple screwed it up when transitioning from "System Preferences" to "System Settings" application and it's still broken. The old "System Preferences" also had both sections (**Added Extensions** and **Finder Extensions**), but FinderSync extensions list always matched in both of them. With "System Settings", **Extensions / Added extensions** is always correct, but **Extensions / Finder extensions** has it's own weird logic, and may not even appear (like you've seen in my example).

So the general rule is, if you want to see icon badges of an extension (GitFinder, or any other), go to **Extensions / Added extensions**, uncheck all **Finder extensions** checkboxes, except for the application whose icon badges you want to see (GitFinder in this case). Once desired badges appear, check back all unchecked extensions, so that you can still see their menu items. Yeah, it's complicated and it sucks, but that's what it is 🙁

> It doesn't create badges (that I know of) but adds entries to the context menu - nicely explained in https://github.com/milke/GitFinder-Issues/issues/123#issuecomment-1763062003.

Yeah, like explained there, once the "winner extension" takes the precedence, it doesn't allow any other extension to display icon badges, even if the "winner extension" doesn't display any badges on its own.

**Comment by Cocoader** [comment]
>  Yeah, like explained there, once the "winner extension" takes the precedence, it doesn't allow any other extension to display icon badges, even if the "winner extension" doesn't display any badges on its own.

Unfortunately the problem remains and returns after the system wakes from sleep. In my eyes this problem can only be _solved by apple_ on the OS level and the problem just surfaced with Hookmark and GitFinder in my case. There are many more combinations that can trigger this behaviour as well hence there is no easy fix in the application.

 I don't get my hopes up that apple will fix this anytime soon so I did read up on [App Extension Programming Guide](https://developer.apple.com/library/archive/documentation/General/Conceptual/ExtensibilityPG/index.html#//apple_ref/doc/uid/TP40014214-CH20-SW1) and found the utility **pluginkit**. 

Pluginkit in combination with an application that can run scripts that can be trigged by something like _system wake_ helped me to _"fix"_ this. I used **Keyboard Maestro** for this because I automated a couple of other things with that already. In case someone has this problem as well here is my macro as a starting point - delays might need to be tighter / looser for your system:
![Image 26 06 24 at 17 13](https://github.com/milke/GitFinder-Issues/assets/13892544/81c09284-cc67-4ac0-be86-4f5219914ac9)
I did test it for the past two weeks and would call it somewhat solved until apple fixes it properly.

**Comment by milke** [comment]
@Cocoader, thanks a lot for this detailed info about how to use **pluginkit** tool to solve the problem. I've already suggested using it in some responses to the problem, reported to me via support email. But I've never made any official suggestion about it, because each customer's case is a little bit different and I was afraid that by generalizing the story I would just make afraid those with simpler cases of the problem. Besides, affected users might have their own preferences of automation tools for the job (you used **Keyboard Maestro**, others may prefer **cron** and shell scripting etc…). But since the problem keeps popping up for years, I assume I will have to do something about proposing such general solution.

**pluginkit** is included in clean macOS installation, so the Installation of **Xcode Command Line Tools** isn't requited (and even if it was, an installation popup would appear the first time the usage of **pluginkit** is attempted). That makes the whole thing easier, it isn't necessary to install eventually unwanted stuff in order to use **pluginkit** to solve the problem.

**Comment by milke** [comment]
> That is good to know for all the git users that don't need Xcode Command Line Tools for their regular work.

GitFinder uses underlying [libgit2](https://libgit2.org) library in pure C, no other tools are needed (installed by macOS installation or otherwise).

**Comment by Cocoader** [comment]
>GitFinder uses underlying libgit2 library in pure C, no other tools are needed (installed by macOS installation or otherwise).

Thank you for the insight. I figured something like that because I looked up your FAQs (_Does GitFinder work well with other git clients?_) before choosing GitFinder.

>That is good to know for all the git users that don't need Xcode Command Line Tools for their regular work.

I might have worded this a bit poorly. What I actually meant to say was that for developer (python, ruby, web development ...) that don't use Xcode directly or parts of that tool chain, don't have to install Xcode Command Line Tools just to use **pluginkit** to fix the extension issue.

**Comment by champagnealexandre** [comment]
Hi! Found this discussion trying to fix the same issue.

In my case, it was BetterTouchTools. Disabling the BTT Finder extension immediately made the badges appear!

<img width="189" alt="image" src="https://github.com/user-attachments/assets/8743986a-609d-4ade-91e2-359ebe049f6c">
